### PR TITLE
Fix: bug when addon dependent an addon in other registry 

### PR DESCRIPTION
**Pull request by wangyikewxgm**
1. fix bug  when addon dependent an addon in other registry 
2. error message when dependency addon not exist 
3. specify an addon in special registry by following command.

```shell
vela addon enable experimental/vector-config
```

Signed-off-by: 楚岳 <wangyike.wyk@alibaba-inc.com>

### Description of your changes

Fixes #

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/addon/addon.go [hunk 1]
```diff
@@ -842,20 +842,23 @@ type Installer struct {
 
 	dryRun     bool
 	dryRunBuff *bytes.Buffer
+
+	dependencyRegistries []Registry
```
Comment (wonderflow) [inline pkg/addon/addon.go:846]:
```suggestion
	registries []Registry
```

File: pkg/addon/addon.go [hunk 2]
```diff
@@ -952,12 +955,30 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
 		if h.dryRun {
 			continue
 		}
-		// always install addon's latest version
+		// try to install the dependent addon from current registry
```
Comment (wonderflow) [inline pkg/addon/addon.go:958]:
```suggestion
		// try to install the dependent addon from the same registry with the current addon
```

File: pkg/addon/addon.go [hunk 3]
```diff
@@ -952,12 +955,30 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
...
+						if errors.Is(err, ErrNotExist) {
+							continue
+						} else {
+							return err
+						}
+					}
+					break
+				}
+				return fmt.Errorf("dependency addon: %s with version: %s cannot be found from all registries", dep.Name, dep.Version)
+			}
```
Comment (wonderflow) [inline pkg/addon/addon.go:979]: image

File: pkg/addon/addon.go [hunk 4]
```diff
@@ -843,22 +843,22 @@ type Installer struct {
 	dryRun     bool
 	dryRunBuff *bytes.Buffer
 
-	dependencyRegistries []Registry
+	Registries []Registry
```
Comment (wonderflow) [inline pkg/addon/addon.go:846]:
```suggestion
	registries []Registry
```

**Review by wonderflow — APPROVED** [review]

File: references/cli/addon.go [hunk 5]
```diff
@@ -1136,3 +1142,15 @@ func NewAddonPackageCommand(c common.Args) *cobra.Command {
...
+
+func checkSpecifyRegistry(name string) (string, string, error) {
+	res := strings.Split(name, "/")
+	switch len(res) {
+	case 2:
+		return res[0], res[1], nil
+	case 1:
+		return "", res[0], nil
+	default:
+		return "", "", fmt.Errorf("error addon name format, you can specify the addon with {addonName} or {addonRegistryName/addonName}")
```
Comment (wonderflow) [inline references/cli/addon.go:1154]:
```suggestion
		return "", "", fmt.Errorf("invalid addon name, you should specify name only  <addonName>  or with registry as prefix <RegistryName>/<addonName>")
```

File: references/cli/addon.go [hunk 6]
```diff
@@ -1136,3 +1142,15 @@ func NewAddonPackageCommand(c common.Args) *cobra.Command {
 	}
 	return cmd
 }
+
+func checkSpecifyRegistry(name string) (string, string, error) {
```
Comment (wonderflow) [inline references/cli/addon.go:1146]:
```suggestion
func splitRegistryAndName(name string) (string, string, error) {
```

File: references/cli/addon.go [hunk 7]
```diff
@@ -543,10 +543,16 @@ func enableAddon(ctx context.Context, k8sClient client.Client, dc *discovery.Dis
 	if err != nil {
 		return err
 	}
-
-	for _, registry := range registries {
+	r, addonName, err := checkSpecifyRegistry(name)
```
Comment (wonderflow) [inline references/cli/addon.go:546]:
```suggestion
	registryName, addonName, err := checkSpecifyRegistry(name)
```

File: references/cli/addon.go [hunk 8]
```diff
@@ -543,10 +543,16 @@ func enableAddon(ctx context.Context, k8sClient client.Client, dc *discovery.Dis
...
+	if err != nil {
+		return err
+	}
```
Comment (wonderflow) [inline references/cli/addon.go:549]: check the registry existence first here

File: references/cli/addon.go [hunk 9]
```diff
@@ -558,21 +564,21 @@ func enableAddon(ctx context.Context, k8sClient client.Client, dc *discovery.Dis
...
 			return err
 		}
-		if err = waitApplicationRunning(k8sClient, name); err != nil {
+		if err = waitApplicationRunning(k8sClient, addonName); err != nil {
 			return err
 		}
 		return nil
 	}
-	return fmt.Errorf("addon: %s not found in registries", name)
+	return fmt.Errorf("addon: %s not found in registries", addonName)
```
Comment (wonderflow) [inline references/cli/addon.go:581]: We should provide different error message according to the registry.

If specified, report:
```
fmt.Errorf("addon: %s not found in registry %s", addonName, registrName)
```

If not specified, report:
```
fmt.Errorf("addon: %s not found in all candidate registries", addonName)
```
 addon: %s not found in registries

**Review by wonderflow — CHANGES_REQUESTED** [review]
still lots changes request

**Comment by wonderflow** [comment]
Please also provide the usage in the PR body.

File: references/cli/addon.go [hunk 10]
```diff
@@ -543,13 +545,24 @@ func enableAddon(ctx context.Context, k8sClient client.Client, dc *discovery.Dis
...
 	}
+	if len(registryName) != 0 {
+		foundRegistry := false
+		for _, registry := range registries {
+			if registry.Name == registryName {
+				foundRegistry = true
+			}
+		}
+		if !foundRegistry {
+			return fmt.Errorf("specified registy not exist")
```
Comment (wonderflow) [inline references/cli/addon.go]:
```suggestion
			return fmt.Errorf("specified registy %s not exist", registryName)
```

**Review by wonderflow — APPROVED** [review]

**Comment by wonderflow** [comment]
In the following, can we still install local addon?

```
vela addon enable addons/pyroscope
```

If not, we should check if it's a local folder first.

**Comment by wangyikewxgm** [comment]
> In the following, can we still install local addon?
> 
> ```
> vela addon enable addons/pyroscope
> ```
> 
> If not, we should check if it's a local folder first.



> In the following, can we still install local addon?
> 
> ```
> vela addon enable addons/pyroscope
> ```
> 
> If not, we should check if it's a local folder first.

Yes, still works.

**Review by Somefive — APPROVED** [review]

**Review by FogDong — APPROVED** [review]

File: pkg/addon/addon.go [hunk 11]
```diff
@@ -952,16 +955,41 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
...
+		if err == nil {
+			if err = depHandler.enableAddon(depAddon); err != nil {
+				return errors.Wrap(err, "fail to dispatch dependent addon resource")
+			}
+			return nil
+		}
+		if !errors.Is(err, ErrNotExist) {
+			return err
+		}
+		for _, registry := range h.registries {
```
Comment (barnettZQG) [inline pkg/addon/addon.go:972]: Retry all registries? The same registry should be excluded.

File: pkg/addon/addon.go [hunk 12]
```diff
@@ -952,16 +955,41 @@ func (h *Installer) installDependency(addon *InstallPackage) error {
...
-		if err != nil {
-			return err
-		}
 		depHandler := *h
 		depHandler.args = nil
-		if err = depHandler.enableAddon(depAddon); err != nil {
-			return errors.Wrap(err, "fail to dispatch dependent addon resource")
+		var depAddon *InstallPackage
+		// try to install the dependent addon from the same registry with the current addon
+		depAddon, err = h.loadInstallPackage(dep.Name, dep.Version)
```
Comment (barnettZQG) [inline pkg/addon/addon.go:962]: Do we support specifying the registry of the dependency?

**Review by barnettZQG — CHANGES_REQUESTED** [review]
The code needs to optimize.

File: pkg/apiserver/domain/service/addon.go [hunk 13]
```diff
@@ -400,8 +400,8 @@ func (u *addonServiceImpl) EnableAddon(ctx context.Context, name string, args ap
 	if err != nil {
 		return err
 	}
-	for _, r := range registries {
-		err = pkgaddon.EnableAddon(ctx, name, args.Version, u.kubeClient, u.discoveryClient, u.apply, u.config, r, args.Args, u.addonRegistryCache)
+	for i, r := range registries {
```
Comment (barnettZQG) [inline pkg/apiserver/domain/service/addon.go:403]: Do we support enabling the addon from a specific registry? I noticed that the CLI supports it.

Comment (wangyikewxgm) [reply pkg/apiserver/domain/service/addon.go:403]: This pr does this thing.

File: pkg/addon/addon.go (hunk 12 above)
Comment (wangyikewxgm) [reply pkg/addon/addon.go:962]: Not yet.

File: pkg/addon/addon.go (hunk 11 above)
Comment (wangyikewxgm) [reply pkg/addon/addon.go:972]: Has excluded.

File: pkg/apiserver/domain/service/addon.go (hunk 13 above)
Comment (barnettZQG) [reply pkg/apiserver/domain/service/addon.go:403]: But, here doesn't support it. We could add a param named registry in EnableAddonRequest.

Comment (wangyikewxgm) [reply pkg/apiserver/domain/service/addon.go:403]: Goted. Will add a parameter in apiserver.

Comment (wangyikewxgm) [reply pkg/apiserver/domain/service/addon.go:414]: Added.

**Comment by wangyikewxgm** [comment]
ping @barnettZQG PLTAL

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**